File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Lang/FrenchLanguagePack.swift
import Foundation

public struct FrenchLanguagePack: LanguagePack {
    public let id = "fr"
    public let displayName = "Français"
    public let locale = Locale(identifier: "fr")

    public init() {}

    public var alphaRows: [[KeyDef]] {
        [
            chars("a", "z", "e", "r", "t", "y", "u", "i", "o", "p"),
            chars("q", "s", "d", "f", "g", "h", "j", "k", "l", "m"),
            [KeyDef("⬆", SpecialKeys.shift, widthWeight: 1.5)]
                + chars("w", "x", "c", "v", "b", "n")
                + [KeyDef("←", SpecialKeys.backspace, widthWeight: 1.5)],
            [
                KeyDef("?123", SpecialKeys.symbols, widthWeight: 1.5),
                KeyDef("🌐", SpecialKeys.globe, widthWeight: 1.0),
                KeyDef(",", keyCode(","), widthWeight: 1.0),
                KeyDef(" ", keyCode(" "), widthWeight: 4.0),
                KeyDef(".", keyCode("."), widthWeight: 1.0),
                KeyDef("↵", SpecialKeys.enter, widthWeight: 1.5),
            ],
        ]
    }

    public var symbols1Rows: [[KeyDef]] { EnglishLanguagePack().symbols1Rows }
    public var symbols2Rows: [[KeyDef]] { EnglishLanguagePack().symbols2Rows }

    public let longPressAccents: [Character: [Character]] = [
        "a": ["à", "â", "ä", "á", "ã"],
        "e": ["é", "è", "ê", "ë"],
        "i": ["î", "ï", "í", "ì"],
        "o": ["ô", "ö", "ó", "ò", "õ"],
        "u": ["ù", "û", "ü", "ú"],
        "c": ["ç"],
        "y": ["ÿ"],
    ]

    public static var appGroupContainer: URL?
    private static let wordlistPackPrefix = "draftright-wordlist-fr"

    public func makeCandidateEngine() -> CandidateEngine? {
        let wordList = WordListPackResolver.loadOrFallback(
            appGroupContainer: Self.appGroupContainer,
            packIdPrefix: Self.wordlistPackPrefix,
            fallback: { FrenchBootstrapWordList.wordList }
        )
        return TrigramCandidateEngine(wordList: wordList)
    }
}

File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/lang/FrenchLanguagePack.kt
package com.draftright.keyboard.lang

import com.draftright.draftright_mobile.v2.R
import com.draftright.keyboard.KeyDef
import com.draftright.keyboard.LanguagePack
import com.draftright.keyboard.SpecialKeys
import com.draftright.keyboard.ime.CandidateEngine
import com.draftright.keyboard.ime.ImeContext
import com.draftright.keyboard.ime.TrigramCandidateEngine
import com.draftright.keyboard.ime.WordListPackResolver
import java.util.Locale

private fun chars(vararg labels: String): List<KeyDef> =
    labels.map { KeyDef(it, it[0].code) }

object FrenchLanguagePack : LanguagePack {
    override val id = "fr"
    override val displayName = "Français"
    override val locale: Locale = Locale.FRENCH

    override val alphaRows: List<List<KeyDef>> = listOf(
        chars("a", "z", "e", "r", "t", "y", "u", "i", "o", "p"),
        chars("q", "s", "d", "f", "g", "h", "j", "k", "l", "m"),
        listOf(KeyDef("⬆", SpecialKeys.SHIFT, 1.5f)) +
            chars("w", "x", "c", "v", "b", "n") +
            listOf(KeyDef("←", SpecialKeys.BACKSPACE, 1.5f)),
        listOf(
            KeyDef("?123", SpecialKeys.SYMBOLS, 1.5f),
            KeyDef("🌐", SpecialKeys.GLOBE, 1.0f),
            KeyDef(",", ','.code, 1.0f),
            KeyDef(" ", ' '.code, 4.0f),
            KeyDef(".", '.'.code, 1.0f),
            KeyDef("↵", SpecialKeys.ENTER, 1.5f),
        ),
    )

    override val symbols1Rows: List<List<KeyDef>> = EnglishLanguagePack.symbols1Rows
    override val symbols2Rows: List<List<KeyDef>> = EnglishLanguagePack.symbols2Rows

    override val longPressAccents: Map<Char, List<Char>> = mapOf(
        'a' to listOf('à', 'â', 'ä', 'á', 'ã'),
        'e' to listOf('é', 'è', 'ê', 'ë'),
        'i' to listOf('î', 'ï', 'í', 'ì'),
        'o' to listOf('ô', 'ö', 'ó', 'ò', 'õ'),
        'u' to listOf('ù', 'û', 'ü', 'ú'),
        'c' to listOf('ç'),
        'y' to listOf('ÿ'),
    )

    @Volatile
    private var cachedEngine: CandidateEngine? = null
    private const val WORDLIST_PACK_PREFIX = "draftright-wordlist-fr"

    override fun candidateEngine(): CandidateEngine? {
        cachedEngine?.let { return it }
        val ctx = ImeContext.appOrNull() ?: return null
        synchronized(this) {
            cachedEngine?.let { return it }
            val list = WordListPackResolver.loadOrFallback(
                context = ctx,
                packIdPrefix = WORDLIST_PACK_PREFIX,
                fallbackResId = R.raw.wordlist_fr,
            )
            val engine = TrigramCandidateEngine(list)
            cachedEngine = engine
            return engine
        }
    }
}
